### PR TITLE
Sonar pulse back end

### PR DESCRIPTION
**Pull request by kramerje**
Implemented the back end functionality for sonar pulse mode.

### CODE REVIEW

File: src/main/java/assets/game.js [hunk 1]
```diff
@@ -178,8 +179,47 @@ function redrawGrid() {
...
+////        document.getElementById("opponent").rows[square.row-1].cells[square.column.charCodeAt(0) - 'A'.charCodeAt(0)].classList.add("occupied");
+//    }));
+
+    // draw in sonar squares from sonar square array
+    game.opponentsBoard.sonarSquares.forEach((square) => {
+//        console.log(square);
+
+        document.getElementById("opponent").rows[square.row-1].cells[square.column.charCodeAt(0) - 'A'.charCodeAt(0)].classList.add("scanned");
+    });
+
```
Comment (aalokb) [inline src/main/java/assets/game.js]: Interesting choice to do this on the back-end, did you make sure to update the application controller to scale with any other potential changes that they might introduce in the next sprint?

File: src/main/java/assets/game.js [hunk 2]
```diff
@@ -219,7 +259,21 @@ function cellClick() {
                 registerCellListener((e) => {});
             }
         });
-    } else {
+    }
+
+    // if isSonar variable is triggered (will be triggered in sonar_attack func, when sonar mode is activated)
+    else if (isSonar) {
```
Comment (aalokb) [inline src/main/java/assets/game.js]: This is good.

File: src/main/java/conf/Routes.java [hunk 3]
```diff
@@ -15,6 +15,7 @@ public void init(Router router) {
         router.GET().route("/game").with(ApplicationController::newGame);
         router.POST().route("/place").with(ApplicationController::placeShip);
         router.POST().route("/attack").with(ApplicationController::attack);
+        router.POST().route("/sonar").with(ApplicationController::sonar_attack); // added sonar_attack function from game.java
```
Comment (aalokb) [inline src/main/java/conf/Routes.java]: Okay, awesome. This should work for the application controller, just make sure that the data being passed between the backend and frontend stays stateless, we dont want the changes on the game.js side to simply just stay on that side, to dissapear when the game object is regenerated....

File: src/main/java/controllers/SonarPulseGameAction.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
+package controllers;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import cs361.battleships.models.Game;
+
+public class SonarPulseGameAction {
```
Comment (aalokb) [inline src/main/java/controllers/SonarPulseGameAction.java]: This should work well, just make sure to keep the @JSONproperty consistent with the rest of the code.

Comment (kramerje) [reply src/main/java/controllers/SonarPulseGameAction.java]: Sounds good, I used the AttackGameAction controller as a template for this new controller

File: src/main/java/conf/Routes.java (hunk 3 above)
Comment (kramerje) [reply src/main/java/conf/Routes.java]: Thanks for the comment, I remember Caius talking about keeping the data stateless. Is there a way to test if the data is in fact stateless? I think did this part correctly?

File: src/main/java/assets/game.js [hunk 5]
```diff
@@ -219,7 +259,21 @@ function cellClick() {
...
+        sendXhr("POST", "/sonar", {game: game, x: row, y: col}, function(data) {
+            // reset isSonar variable to false again (so in normal attack mode)
+            isSonar = false;
```
[the remark this answers is not in the document]
Comment (kramerje) [reply src/main/java/assets/game.js]: You're right, I think that's a bug or feature we need to address next.

**Pull request merged**